# **Apple App Data**

Complate Explotry Data Analysis

## About Dataset

**Context**

Apple AppStore Android App Data. (1.2 Million+ App Data)

**Content**

I've collected the data with the help of Python script (Scrapy) running on a cluster of cloud vm instances.
The data was collected in the month of October 2021.


>- **`Description`**\
> The Data Set was downloaded from Kaggle, from the following[Link](https://www.kaggle.com/datasets/gauthamp10/apple-appstore-apps/data?select=appleAppData.csv)


In [1]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#this is for jupyter notebook to show the plot in the notebook itself instead of opening a new window for the plot
%matplotlib inline 

In [2]:
import warnings
warnings.filterwarnings('ignore')


## 2. **Data Loading and exploration and cleaning**
 ↪ Load the csv file with the pandas
 
 ↪ creating the dataframe and understanding the data present in the dataset using pandas
 
 ↪ Dealing with the missing data, outliers and the incorrect records

In [3]:
df = pd.read_csv('./Data/AppleAppData.csv')


In [4]:
pd.set_option('display.max_columns', None) # this is to display all the columns in the dataframe
pd.set_option('display.max_rows', None) # this is to display all the rows in the dataframe

In [5]:
df.head(5)

App_Id                 App_Name  \
0               com.hkbu.arc.apaper           A+ Paper Guide   
1               com.dmitriev.abooks                  A-Books   
2                    no.terp.abooks                  A-books   
3          fr.antoinettefleur.Book1              A-F Book #1   
4  com.imonstersoft.azdictionaryios  A-Z Synonyms Dictionary   

                                        AppStore_Url Primary_Genre  \
0  https://apps.apple.com/us/app/a-paper-guide/id...     Education   
1  https://apps.apple.com/us/app/a-books/id103157...          Book   
2  https://apps.apple.com/us/app/a-books/id145702...          Book   
3  https://apps.apple.com/us/app/a-f-book-1/id500...          Book   
4  https://apps.apple.com/us/app/a-z-synonyms-dic...     Reference   

  Content_Rating  Size_Bytes Required_IOS_Version              Released  \
0             4+  21993472.0                  8.0  2017-09-28T03:02:41Z   
1             4+  13135872.0                 10.0  2015-08-31T19:31:32Z   
2             4+  21943296.0                  9.0  2021-04-14T07:00:00Z   
3             4+  81851392.0                  8.0  2012-02-10T03:40:07Z   
4             4+  64692224.0                  9.0  2020-12-16T08:00:00Z   

                Updated Version  Price Currency   Free  DeveloperId  \
0  2018-12-21T21:30:36Z   1.1.2   0.00      USD   True   1375410542   
1  2019-07-23T20:31:09Z     1.3   0.00      USD   True   1031572001   
2  2021-05-30T21:08:54Z   1.3.1   0.00      USD   True   1457024163   
3  2019-10-29T12:40:37Z     1.2   2.99      USD  False    439568839   
4  2020-12-18T21:36:11Z   1.0.1   0.00      USD   True    656731821   

        Developer                                      Developer_Url  \
0        HKBU ARC  https://apps.apple.com/us/developer/hkbu-arc/i...   
1  Roman Dmitriev  https://apps.apple.com/us/developer/roman-dmit...   
2         Terp AS  https://apps.apple.com/us/developer/terp-as/id...   
3   i-editeur.com  https://apps.apple.com/us/developer/i-editeur-...   
4   Ngov chiheang  https://apps.apple.com/us/developer/ngov-chihe...   

         Developer_Website  Average_User_Rating  Reviews  \
0                      NaN                  0.0        0   
1                      NaN                  5.0        1   
2                      NaN                  0.0        0   
3                      NaN                  0.0        0   
4  http://imonstersoft.com                  0.0        0   

   Current_Version_Score  Current_Version_Reviews  
0                    0.0                        0  
1                    5.0                        1  
2                    0.0                        0  
3                    0.0                        0  
4                    0.0                        0

In [6]:
df.tail(5)

App_Id                    App_Name  \
1230371               com.ledtech.sadblock          Sесurity АdBlосkеr   
1230372                    com.securex.vpn    SесurеХ VРN - Wifi Proxy   
1230373            com.beelab.SoTayXayDung             Sổ tay Xây dựng   
1230374                       com.icc.sttb  Sổ tay đảng viên Thái Bình   
1230375  com.vnptlonganios.sodiemthongminh          Sổ Điểm Thông Minh   

                                              AppStore_Url Primary_Genre  \
1230371  https://apps.apple.com/us/app/s%D0%B5%D1%81uri...     Utilities   
1230372  https://apps.apple.com/us/app/s%D0%B5%D1%81ur%...     Utilities   
1230373  https://apps.apple.com/us/app/s%E1%BB%95-tay-x...     Utilities   
1230374  https://apps.apple.com/us/app/s%E1%BB%95-tay-%...     Utilities   
1230375  https://apps.apple.com/us/app/s%E1%BB%95-%C4%9...     Utilities   

        Content_Rating  Size_Bytes Required_IOS_Version              Released  \
1230371             4+  16666624.0                 13.0  2020-07-07T07:00:00Z   
1230372             4+  39016448.0                  9.0  2019-02-12T10:10:13Z   
1230373             4+  17223680.0                  9.0  2018-10-17T04:22:41Z   
1230374             4+  56716288.0                 10.0  2021-02-20T08:00:00Z   
1230375             4+  85135360.0                  8.0  2018-06-05T07:45:41Z   

                      Updated Version  Price Currency  Free  DeveloperId  \
1230371  2020-07-10T00:48:50Z   1.0.1    0.0      USD  True   1522287989   
1230372  2020-10-21T23:25:15Z     1.1    0.0      USD  True   1492288123   
1230373  2018-10-17T04:22:41Z     1.0    0.0      USD  True   1438594214   
1230374  2021-10-02T22:00:19Z   1.2.5    0.0      USD  True   1515469508   
1230375  2019-05-21T22:03:17Z     1.9    0.0      USD  True   1350355912   

                Developer                                      Developer_Url  \
1230371  LED-TECHNOLOGIES  https://apps.apple.com/us/developer/led-techno...   
1230372    Trust VPN Ltd.  https://apps.apple.com/us/developer/trust-vpn-...   
1230373          Luu Minh  https://apps.apple.com/us/developer/luu-minh/i...   
1230374         Thái Bình  https://apps.apple.com/us/developer/th%C3%A1i-...   
1230375     Pham Thanh Vo  https://apps.apple.com/us/developer/pham-thanh...   

                 Developer_Website  Average_User_Rating  Reviews  \
1230371                        NaN              3.91608      143   
1230372    https://securexvpn.com/              4.82733     1500   
1230373  http://bee-labs.github.io              4.00000        1   
1230374       https://aisoftech.vn              0.00000        0   
1230375                        NaN              0.00000        0   

         Current_Version_Score  Current_Version_Reviews  
1230371                3.91608                      143  
1230372                4.82733                     1500  
1230373                4.00000                        1  
1230374                0.00000                        0  
1230375                0.00000                        0

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1230376 entries, 0 to 1230375
Data columns (total 21 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   App_Id                   1230376 non-null  object 
 1   App_Name                 1230375 non-null  object 
 2   AppStore_Url             1230376 non-null  object 
 3   Primary_Genre            1230376 non-null  object 
 4   Content_Rating           1230376 non-null  object 
 5   Size_Bytes               1230152 non-null  float64
 6   Required_IOS_Version     1230376 non-null  object 
 7   Released                 1230373 non-null  object 
 8   Updated                  1230376 non-null  object 
 9   Version                  1230376 non-null  object 
 10  Price                    1229886 non-null  float64
 11  Currency                 1230376 non-null  object 
 12  Free                     1230376 non-null  bool   
 13  DeveloperId              1230376 non-null 

In [8]:
df.describe() #this will give the summary of the data

Size_Bytes         Price   DeveloperId  Average_User_Rating  \
count  1.230152e+06  1.229886e+06  1.230376e+06         1.230376e+06   
mean   7.510857e+07  5.041264e-01  1.092869e+09         1.788387e+00   
std    2.318726e+08  6.869674e+00  4.090897e+08         2.154105e+00   
min    2.762800e+04  0.000000e+00  2.816565e+08         0.000000e+00   
25%    2.041344e+07  0.000000e+00  7.381109e+08         0.000000e+00   
50%    4.333158e+07  0.000000e+00  1.179711e+09         0.000000e+00   
75%    8.365466e+07  0.000000e+00  1.471960e+09         4.333330e+00   
max    7.677945e+10  9.999900e+02  1.590233e+09         5.000000e+00   

            Reviews  Current_Version_Score  Current_Version_Reviews  
count  1.230376e+06           1.230376e+06             1.230376e+06  
mean   8.259478e+02           1.788387e+00             8.259478e+02  
std    5.016051e+04           2.154105e+00             5.016051e+04  
min    0.000000e+00           0.000000e+00             0.000000e+00  
25%    0.000000e+00           0.000000e+00             0.000000e+00  
50%    0.000000e+00           0.000000e+00             0.000000e+00  
75%    3.000000e+00           4.333330e+00             3.000000e+00  
max    2.268533e+07           5.000000e+00             2.268533e+07

In [9]:
df.shape

(1230376, 21)

In [10]:
df.columns
#this will give the columns names


Index(['App_Id', 'App_Name', 'AppStore_Url', 'Primary_Genre', 'Content_Rating',
       'Size_Bytes', 'Required_IOS_Version', 'Released', 'Updated', 'Version',
       'Price', 'Currency', 'Free', 'DeveloperId', 'Developer',
       'Developer_Url', 'Developer_Website', 'Average_User_Rating', 'Reviews',
       'Current_Version_Score', 'Current_Version_Reviews'],
      dtype='object')

In [11]:
df.isnull().sum() #this will give the number of null values in each column

App_Id                          0
App_Name                        1
AppStore_Url                    0
Primary_Genre                   0
Content_Rating                  0
Size_Bytes                    224
Required_IOS_Version            0
Released                        3
Updated                         0
Version                         0
Price                         490
Currency                        0
Free                            0
DeveloperId                     0
Developer                       0
Developer_Url                1109
Developer_Website          643988
Average_User_Rating             0
Reviews                         0
Current_Version_Score           0
Current_Version_Reviews         0
dtype: int64

In [12]:
#this will give the percentage of null values in each column
(df.isnull().sum() / len(df) * 100).sort_values(ascending=False)

Developer_Website          52.340748
Developer_Url               0.090135
Price                       0.039825
Size_Bytes                  0.018206
Released                    0.000244
App_Name                    0.000081
Primary_Genre               0.000000
AppStore_Url                0.000000
App_Id                      0.000000
Content_Rating              0.000000
Version                     0.000000
Updated                     0.000000
Required_IOS_Version        0.000000
Free                        0.000000
Currency                    0.000000
Developer                   0.000000
DeveloperId                 0.000000
Average_User_Rating         0.000000
Reviews                     0.000000
Current_Version_Score       0.000000
Current_Version_Reviews     0.000000
dtype: float64

In [13]:
df['Content_Rating'].value_counts() 

Content_Rating
4+               980971
17+              124127
12+               89570
9+                35698
Not yet rated        10
Name: count, dtype: int64

In [14]:
df['Content_Rating'].unique() #this will give the unique values in the column

array(['4+', '17+', '9+', '12+', 'Not yet rated'], dtype=object)

In [15]:
df['Required_IOS_Version'].value_counts() #this will give the number of apps in each os version

Required_IOS_Version
11.0       210764
9.0        206440
10.0       196082
8.0        150973
12.0        94489
13.0        52161
7.0         46953
6.0         42845
12.1        20518
9.3         16151
10.3        15753
14.0        12660
12.2        12158
7.1          9520
8.1          7726
11.4         7329
12.4         6378
11.2         6347
9.1          6008
14.1         6007
5.1.1        5903
13.2         5673
10.2         5531
11.1         4876
6.1          4312
4.3          4177
10.1         3954
11.3         3912
8.2          3813
9.2          3280
13.4         2871
12.3         2575
14.5         2553
5.0          2542
13.6         2328
14.4         2300
10.7         2216
10.10        2183
10.11        2167
13.1         2048
8.4          1772
13.5         1756
10.12        1729
10.9         1637
11           1602
14.3         1474
10.9.0       1435
10.13        1321
10.15        1300
14.2         1260
10           1256
10.8         1252
10.14        1189
10.6         1185
10.6.6 

In [16]:
df['Required_IOS_Version'].unique() #this will give the unique values in the column

array(['8.0', '10.0', '9.0', '7.1', '12.0', '11.0', '13.0', '7.0', '9.3',
       '5.1.1', '11.3', '6.0', '12.1', '8.2', '6.1', '11.4', '10.3',
       '9.1', '8.4', '9.2', '4.0', '5.1', '8.1', '5.0', '10.10.0', '14.5',
       '14.0', '14.1', '4.3', '13.5', '11', '12.4', '10.10', '11.2',
       '14.4', '11.1', '12.2', '5.2', '10.1', '14.3', '4.2', '13.2',
       '3.2', '8.3', '3.1.3', '10.2', '10.9.0', '13.6', '3.0', '13.3',
       '9', '11.0.0', '3.1.2', '14.2', '13.4', '12.3', '13.1', '12',
       '9.3.5', '10', '9.4', '7.2', '9.0.0', '10.12', '13', '7', '10.13',
       '3.1', '2.2.1', '10.9', '10.8', '10.7', '12.0.0', '10.15', '10.5',
       '10.14', '15.0', '13.7', '10.11', '4.1', '10.6', '10.6.6', '7.0.0',
       '14.7', '10.13.0', '10.9.5', '12.5', '8', '6', '10.0.0', '10.6.8',
       '7.1.2', '10.12.2', '10.11.0', '14', '10.7.4', '10.3.3', '12.4.4',
       '10.8.0', '10.14.6', '10.4.0', '8.0.0', '10.15.1', '10.14.2',
       '10.15.0', '10.13.6', '14.0.0', '10.13.2', '10.7.3', '10.

In [17]:
df['Version'].value_counts() #this will give the number of apps in each version

Version
1.0                                                                218134
1.1                                                                 71760
1.2                                                                 43369
1.0.1                                                               31051
2.0                                                                 27542
1.3                                                                 27418
1.0.2                                                               20330
1.0.0                                                               19507
1.4                                                                 18554
1.0.3                                                               14507
1.5                                                                 14354
2.1                                                                 11593
1.1.0                                                               10652
1.0.4                         

In [18]:
df['Version'].unique() #this will give the unique values in the column

array(['1.1.2', '1.3', '1.3.1', ..., '1.40.0.1', '10.2a.0001',
       '2.2021.09'], shape=(41270,), dtype=object)

# **Observations**
---
1. There are 1230376 rows and 21 columns in the dataset
2. The columns are of different data types
3. The columns in the datasets are:
   - `'App_Id', 'App_Name', 'AppStore_Url', 'Primary_Genre', 'Content_Rating',
       'Size_Bytes', 'Required_IOS_Version', 'Released', 'Updated', 'Version',
       'Price', 'Currency', 'Free', 'DeveloperId', 'Developer',
       'Developer_Url', 'Developer_Website', 'Average_User_Rating', 'Reviews',
       'Current_Version_Score', 'Current_Version_Reviews
4. There are some missing values in the dataset which we will read in details and deal later on in the notebook.
5. There are some columns which are of object data type but they should be of numeric data type, we will convert them later on in the notebook.
   - `'Required_IOS_Version', 'Version'` 

In [19]:
df.describe()

Size_Bytes         Price   DeveloperId  Average_User_Rating  \
count  1.230152e+06  1.229886e+06  1.230376e+06         1.230376e+06   
mean   7.510857e+07  5.041264e-01  1.092869e+09         1.788387e+00   
std    2.318726e+08  6.869674e+00  4.090897e+08         2.154105e+00   
min    2.762800e+04  0.000000e+00  2.816565e+08         0.000000e+00   
25%    2.041344e+07  0.000000e+00  7.381109e+08         0.000000e+00   
50%    4.333158e+07  0.000000e+00  1.179711e+09         0.000000e+00   
75%    8.365466e+07  0.000000e+00  1.471960e+09         4.333330e+00   
max    7.677945e+10  9.999900e+02  1.590233e+09         5.000000e+00   

            Reviews  Current_Version_Score  Current_Version_Reviews  
count  1.230376e+06           1.230376e+06             1.230376e+06  
mean   8.259478e+02           1.788387e+00             8.259478e+02  
std    5.016051e+04           2.154105e+00             5.016051e+04  
min    0.000000e+00           0.000000e+00             0.000000e+00  
25%    0.000000e+00           0.000000e+00             0.000000e+00  
50%    0.000000e+00           0.000000e+00             0.000000e+00  
75%    3.000000e+00           4.333330e+00             3.000000e+00  
max    2.268533e+07           5.000000e+00             2.268533e+07

## Observations:
---
- We have only 7 columns as numeric data type, rest all are object data type (according to python), but we can see that `'Required_IOS_Version', 'Version'` are also numeric, we must convert them to numeric data type in data wrangling process.
---

## Let's Clean the 'Content_Rating' column First

In [20]:
df['Content_Rating'].value_counts()

Content_Rating
4+               980971
17+              124127
12+               89570
9+                35698
Not yet rated        10
Name: count, dtype: int64

In [21]:
df['Content_Rating'].isnull().sum()

np.int64(0)

No null Value,Its good

In [22]:
df['Content_Rating'].unique()

array(['4+', '17+', '9+', '12+', 'Not yet rated'], dtype=object)

## Lets Clean the Required_IOS_Version	

In [23]:
df['Required_IOS_Version'].unique()

array(['8.0', '10.0', '9.0', '7.1', '12.0', '11.0', '13.0', '7.0', '9.3',
       '5.1.1', '11.3', '6.0', '12.1', '8.2', '6.1', '11.4', '10.3',
       '9.1', '8.4', '9.2', '4.0', '5.1', '8.1', '5.0', '10.10.0', '14.5',
       '14.0', '14.1', '4.3', '13.5', '11', '12.4', '10.10', '11.2',
       '14.4', '11.1', '12.2', '5.2', '10.1', '14.3', '4.2', '13.2',
       '3.2', '8.3', '3.1.3', '10.2', '10.9.0', '13.6', '3.0', '13.3',
       '9', '11.0.0', '3.1.2', '14.2', '13.4', '12.3', '13.1', '12',
       '9.3.5', '10', '9.4', '7.2', '9.0.0', '10.12', '13', '7', '10.13',
       '3.1', '2.2.1', '10.9', '10.8', '10.7', '12.0.0', '10.15', '10.5',
       '10.14', '15.0', '13.7', '10.11', '4.1', '10.6', '10.6.6', '7.0.0',
       '14.7', '10.13.0', '10.9.5', '12.5', '8', '6', '10.0.0', '10.6.8',
       '7.1.2', '10.12.2', '10.11.0', '14', '10.7.4', '10.3.3', '12.4.4',
       '10.8.0', '10.14.6', '10.4.0', '8.0.0', '10.15.1', '10.14.2',
       '10.15.0', '10.13.6', '14.0.0', '10.13.2', '10.7.3', '10.

In [24]:
df['Required_IOS_Version'].value_counts()

Required_IOS_Version
11.0       210764
9.0        206440
10.0       196082
8.0        150973
12.0        94489
13.0        52161
7.0         46953
6.0         42845
12.1        20518
9.3         16151
10.3        15753
14.0        12660
12.2        12158
7.1          9520
8.1          7726
11.4         7329
12.4         6378
11.2         6347
9.1          6008
14.1         6007
5.1.1        5903
13.2         5673
10.2         5531
11.1         4876
6.1          4312
4.3          4177
10.1         3954
11.3         3912
8.2          3813
9.2          3280
13.4         2871
12.3         2575
14.5         2553
5.0          2542
13.6         2328
14.4         2300
10.7         2216
10.10        2183
10.11        2167
13.1         2048
8.4          1772
13.5         1756
10.12        1729
10.9         1637
11           1602
14.3         1474
10.9.0       1435
10.13        1321
10.15        1300
14.2         1260
10           1256
10.8         1252
10.14        1189
10.6         1185
10.6.6 

In [25]:
df['Required_IOS_Version'].isnull().sum()
# No Null Value

np.int64(0)

In [26]:
df['Required_IOS_Version'] = df['Required_IOS_Version'].astype(str) #this will convert the column to string
df['Required_IOS_Version'] = df['Required_IOS_Version'].astype(float) #this will convert the column to float


ValueError: could not convert string to float: '5.1.1'